localstack/services/s3/cors.py

```diff
@@ -0,0 +1,250 @@
+import re
+from typing import Optional, Tuple
+
+from werkzeug.datastructures import Headers
+
+from localstack import config
+from localstack.aws.api import RequestContext
+from localstack.aws.api.s3 import AccessForbidden, BadRequest, CORSRule, CORSRules
+from localstack.aws.chain import Handler, HandlerChain
+
+# TODO: refactor those to expose the needed methods
+from localstack.aws.handlers.cors import CorsEnforcer, CorsResponseEnricher
+from localstack.aws.protocol.op_router import RestServiceOperationRouter
+from localstack.aws.protocol.serializer import gen_amzn_requestid_long
+from localstack.aws.protocol.service_router import get_service_catalog
+from localstack.constants import S3_VIRTUAL_HOSTNAME
+from localstack.http import Request, Response
+from localstack.services.s3.models import BucketCorsIndex
+from localstack.services.s3.utils import S3_VIRTUAL_HOSTNAME_REGEX
+
+_s3_virtual_host_regex = re.compile(S3_VIRTUAL_HOSTNAME_REGEX)
+FAKE_HOST_ID = "9Gjjt1m+cjU4OPvX9O9/8RuvnG41MRb/18Oux2o5H5MY7ISNTlXN+Dz9IG62/ILVxhAGI0qyPfg="
+
+# TODO: refactor those to expose the needed methods maybe in another way that both can import
+add_default_headers = CorsResponseEnricher.add_cors_headers
+is_origin_allowed_default = CorsEnforcer.is_cors_origin_allowed
+
+
+class S3CorsHandler(Handler):
+
+    bucket_cors_index: BucketCorsIndex
+
+    def __init__(self):
+        self.bucket_cors_index = BucketCorsIndex()
+        self._service = get_service_catalog().get("s3")
+        self._s3_op_router = RestServiceOperationRouter(self._service)
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        self.handle_cors(chain, context, response)
+
+    def pre_parse_s3_request(self, context: RequestContext) -> Tuple[bool, Optional[str]]:
```

Comment:
Could you please add a docstring here which explains what the function is doing (or at least what it is returning)?